docs/design/proposals/healer-controller.md

```diff
@@ -0,0 +1,115 @@
+* We currently do not have a way to act on failed volumes. Once the controller detects that the volume is not healthy, it will have an opportunity to act on it. For example userspace mounters such as NBD has a process running per volume, this controller should ideally help restart such processes in case if they are not-started or crashed.
+* This will be a generic controller design for any mounters added in the future, for userspace mounters it will offer healing in addition, like restarting the respective processes if the volume goes abnormal. Some example userspace mounters would be rbd-nbd, cephfs-fuse, etc.
+
+# Details
+
+                    .-------------------.
+                    |     CSI-Addons    |
+                    | Healer Controller |
+                    '-------------------'
+                              |
+                              | gRPC
+                              |
+            .---------+------------------------------.
+            |         |                              |
+            |  .------------.        .------------.  |
+            |  | CSI-Addons |  gRPC  |    CSI     |  |
+            |  |  side-car  |--------| NodePlugin |  |
+            |  '------------'        '------------'  |
+            | CSI-driver Pod                         |
+            '----------------------------------------'
+
+## CSI-Addons Healer Controller
+* The healer controller will be implemented as part of the CSI-Addons controller. It will trigger controller RPC to check the health condition of the CSI volumes.
+* Watch for the `CSIAddonsNode` object, if the object is created/modified by the side-car do below operations
+    * Fetch all the VolumeAttachment` list
+    * Filter the `VolumeAttachment` list through matching driver name, node name and status attached
+    * For each `VolumeAttachment` get the respective PersistentVolume information and check the criteria of PersistentVolume Bound (also we can have configuration options like mounter type)
+    * Send the RPC request for side-car to check if the volume is healthy along with the PersistentVolume details and may attempt to heal the volume if it is found abnormal
+    * Log the final response returned from the side-car
+
+## CSI-Addons side-car
+* When a nodeplugin container is restarted the side-car will update/create(if not present) [CSIAddonsNode](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/api/v1alpha1/csiaddonsnode_types.go) object.
+* The side-car works as relay, upon receiving a gRPC call from healer controller, the CSI-Addons side-car will then check volume's mounting conditions collected via the CSI NodeGetVolumeStats RPC.
+* In case if any volume condition is reported abnormal, the CSI-Addons side-car will act on it, it does some internal operation like fetching secrets, constructing staging path and finally sends StageVolume request and finally returns back the response to the controller.
```

Comment:
We agreed it will be a single call?
and let the driver decide and heal on its own?

Reply:
Yes, for the restart case only, send everything in one go.